source/MaterialXGenOgsXml/OgsXmlGenerator.cpp

```diff
@@ -0,0 +1,243 @@
+            func.append_attribute(FUNCTION_VAL) = functionName.c_str();
+            pugi::xml_node source = impl.append_child(SOURCE);
+            source.append_child(pugi::node_cdata).set_value(functionSource.c_str());
+        }
+    }
+
+    void xmlSetProperty(pugi::xml_node& prop, const string& name, const string& variable, const string& flags = EMPTY_STRING)
+    {
+        prop.append_attribute(NAME) = variable.c_str();
+        auto semantic = OGS_SEMANTICS_MAP.find(name);
+        if (semantic != OGS_SEMANTICS_MAP.end())
+        {
+            prop.append_attribute(SEMANTIC) = semantic->second;
+        }
+        if (!flags.empty())
+        {
+            prop.append_attribute(FLAGS) = flags.c_str();
+        }
+    }
+
+    void xmlAddProperties(pugi::xml_node& parent, const VariableBlock& block, const string& flags = EMPTY_STRING)
+    {
+        for (size_t i = 0; i < block.size(); ++i)
+        {
+            const ShaderPort* p = block[i];
+            if (p->getType() == Type::FILENAME)
+            {
+                const string& samplerName = p->getVariable();
+                if (samplerName.size() > 7 && samplerName.substr(samplerName.size() - 7) == OgsXmlGenerator::SAMPLER_SUFFIX)
+                {
+                    string textureName = samplerName.substr(0, samplerName.size() - 7);
+                    pugi::xml_node texture = parent.append_child(TEXTURE2);
+                    xmlSetProperty(texture, p->getName(), textureName, flags);
+                    pugi::xml_node sampler = parent.append_child(SAMPLER);
+                    xmlSetProperty(sampler, p->getName(), samplerName, flags);
+                }
+            }
+            else
+            {
+                auto type = OGS_TYPE_MAP.find(p->getType());
+                if (type != OGS_TYPE_MAP.end())
+                {
+                    pugi::xml_node prop = parent.append_child(type->second);
+                    xmlSetProperty(prop, p->getName(), p->getVariable(), flags);
+                }
+            }
+        }
+    }
+
+    void xmlAddValues(pugi::xml_node& parent, const VariableBlock& block)
+    {
+        for (size_t i = 0; i < block.size(); ++i)
+        {
+            const ShaderPort* p = block[i];
+            if (p->getValue())
+            {
+                auto type = OGS_TYPE_MAP.find(p->getType());
+                if (type != OGS_TYPE_MAP.end())
+                {
+                    pugi::xml_node val = parent.append_child(type->second);
+                    val.append_attribute(NAME) = p->getVariable().c_str();
+                    val.append_attribute(VALUE) = p->getValue()->getValueString().c_str();
+                }
+            }
+        }
+    }
+}
+
+const string OgsXmlGenerator::SAMPLER_SUFFIX = "Sampler";
+
+OgsXmlGenerator::OgsXmlGenerator()
+{
+}
+
+void OgsXmlGenerator::generate(const Shader* glsl, const Shader* hlsl, std::ostream& stream)
+{
+    if (glsl == nullptr && hlsl == nullptr)
```

Comment:
This will crash if glsl is null and hlsl is not-null. I guess for now we assume glsl must always there. Looks like line 209 should be called to set shader first, then use that instead of glsl-> at line 199 etc.

Reply:
Good catch! Will fix this.